packages/oidc-client/src/server/middleware.ts

```
@@ -0,0 +1,108 @@
+import { NextMiddleware } from "next/server";
+import { NextResponse } from "next/server";
+import {
+  Config,
+  clientSessionPath,
+  internalUnauthorizedPath,
+} from "@/lib/config";
+import { Session } from "@/lib/types";
+import { internalExchangeTokenForSession } from "@lib/core";
+
+type Options = {
+  prepend?: (args: { token: string; userID: string }) => Promise<void> | void;
+  onError?: (err: Error) => Promise<void> | void;
+};
+
+export const withAuth = (
+  config: Config,
+  options?: Options,
+  middlware?: NextMiddleware,
+): NextMiddleware => {
+  return async (request, event) => {
+    const nextURL = request.nextUrl;
+    if (nextURL.pathname.startsWith(config.loginCallbackPath())) {
+      const result = await handleCallback(
+        request.nextUrl.searchParams,
+        config,
+        options,
+      );
+      if (result instanceof NextResponse) {
+        return result;
+      }
+    } else if (nextURL.pathname.startsWith(clientSessionPath)) {
+      const tailorToken = request.cookies.get("tailor.token");
+      return NextResponse.json({ token: tailorToken?.value });
+    } else if (nextURL.pathname.startsWith(internalUnauthorizedPath)) {
+      return NextResponse.redirect(config.appUrl(config.unauthorizedPath()));
+    }
+
+    await middlware?.(request, event);
+  };
+};
+
+export class CallbackError extends Error {
+  constructor(
+    readonly name: string,
+    readonly message: string,
+  ) {
+    super(message);
+  }
+}
+
+export const paramsError = () =>
+  new CallbackError("invalid-params", "code and redirectURI should be filled");
+export const exchangeError = (reason: string) =>
+  new CallbackError("failed-exchange", reason);
+
+export const handleCallback = async (
+  params: URLSearchParams,
+  config: Config,
+  options?: Options,
+) => {
+  try {
+    const code = params.get("code");
+    const redirectURI = params.get("redirect_uri");
+    if (!code || !redirectURI) {
```

for extra security, I think we should throw an error if the URI isn't a path, e.g. someone could highjack this and redirect to a malicious URL.

---

Agreed. I will add additinonal check here.

---

I guess the redirect URI here is always formed as URL, right? 

The URL built at: https://github.com/tailor-platform/frontend-packages/pull/32/files#diff-c3c370e9ed584c0103b358fd1344d0472b3e47202eb604f5508b2320486a80fcR21-R23

---

If so, how about validating if the redirect destination is a path or not in `login` function instead?

https://github.com/tailor-platform/frontend-packages/pull/32/files#diff-c3c370e9ed584c0103b358fd1344d0472b3e47202eb604f5508b2320486a80fcR18-R25

---

nextjs' redirect function accepts paths too so I think we can require a path to be passed. Server side component might need a full URL though...? Can always kick the can for now and add a task to properly design/implement a solid solution.

---

Let's do that. I would like to put off the decision for now.